idseq_dag/util/count.py

```diff
@@ -18,86 +17,40 @@ class ReadCountingMode(Enum):
-        file_format = local_file_path.split(".")[-1]
-
-    named_args = {
-        'local_file_path': local_file_path
-    }
-
-    if max_reads:
-        max_lines = _reads2lines(max_reads, file_format)
-        assert max_lines is not None, "Could not convert max_reads to max_lines"
-        cmd += r''' | head -n "${max_lines}"'''
-        named_args.update({
-            'max_lines': max_lines
-        })
-
-    cmd += " |  wc -l"
-
-    cmd_output = command.execute_with_output(
-        command_patterns.ShellScriptCommand(
-            script=cmd,
-            named_args=named_args
-        )
-    )
-    line_count = int(cmd_output.strip().split(' ')[0])
-    return _lines2reads(line_count, file_format)
-
-
-def _lines2reads(line_count, file_format):
-    '''
-    Convert line count to read count based on file format.
-    Supports fastq and SINGLE-LINE fasta formats.
-    TODO: add support for m8 files here once the relevant steps are added in the pipeline engine.
+def count_reads(filename):
     '''
-    read_count = None
-    if file_format in ["fq", "fastq"]:
-        # Each read consists of exactly 4 lines, by definition.
-        assert line_count % 4 == 0, "File does not follow fastq format"
-        read_count = line_count // 4
-    elif file_format in ["fa", "fasta"]:
-        # Each read consists of exactly 2 lines (identifier and sequence).
-        # ASSUMES the format is SINGLE-LINE fasta files, where each read's sequence
-        # is on a single line rather than being spread over multiple lines.
-        # This is fine for now, because we only count fasta files we generated
-        # ourselves, following the single-line format.
-        assert line_count % 2 == 0, "File does not follow single-line fasta format"
-        read_count = line_count // 2
-    else:
-        read_count = line_count
-    return read_count
-
-
-def _reads2lines(read_count, file_format):
+    Count reads in a given FASTA or FASTQ file.
     '''
-    Convert read count to line count based on file format.
-    Currently supports fastq or SINGLE-LINE fasta.
-    '''
-    line_count = None
-    if file_format in ["fq", "fastq"]:
-        line_count = 4 * read_count
-    elif file_format in ["fa", "fasta"]:
-        # Assumes the format is single-line fasta rather than multiline fasta
-        line_count = 2 * read_count
-    return line_count
+    libmagic_cmd = ["file", "--dereference", "--brief", "--mime-type", filename]
+    is_gzipped = True if run(libmagic_cmd, stdout=PIPE).stdout.startswith(b"application/gzip") else False
+    with gzip.open(filename) if is_gzipped else open(filename, mode="rb") as fmt_fh:
+        first_char = fmt_fh.read(1).decode()[0]
+    with open(filename, "rb") as fh:
+        if first_char == ">":
+            cmd = "grep -c '^>'"
+            if is_gzipped:
+                cmd = "gunzip | " + cmd
+            return int(run(cmd, stdin=fh, stdout=PIPE, check=True, shell=True).stdout)
```

Comment:
Are we moving away from the standardized subprocess module `idseq_dag/util/command.py`? IIRC this was a security recommendation to isolate the risks of constructing arbitrary commands. It also provides logging. If we have decided to move away from it disregard but otherwise we should use it consistently.

Reply:
The commands here do not perform any shell interpolation of strings, and therefore the security recommendation is not applicable.

I have several severe issues with `idseq_dag/util/command.py` and I expect to completely retire that module now that the overall interface for workflow inputs is well-defined. Among the issues:
- Extreme verbosity of logging, to the point where it's impossible to understand the job logs
- Extreme verbosity that this module adds everywhere in the codebase
- High levels of indirection and resulting cyclomatic complexity making it very hard to understand what happens to the command, its arguments, and other parameters that are normally passed to `subprocess.Popen()`
The excessive logging and indirection present in `idseq_dag/util/command.py` and `idseq_dag/util/log.py` are not the solution to our problems. Instead, we should sanitize pipeline input and use standard library components designed for the purpose of logging and subprocess execution. Filename inputs should never be interpolated in shell string input, but can be passed directly to the subprocess as positional command line arguments without using the shell, or on stdin as seen here.

Reply:
I understand the frustration but disagree with the solution of bypassing. 
It might argued that the methods have high complexity and verbosity, but those arguments should be used to simplify those utility functions in terms of api (keep the api) or logging. We should not be bypassing them completely. The addition of those components, allowed us to establish a unique format and level of logs across the whole application and get a clearer picture of what was happening (in particular in case of complex bugs). I find having a single point where we can determine and change over time the level of logging is useful.
The JSON nature of the logs adds to the verbosity, but that can be handled by using the message field for the human readable output.

Reply:
I understand the desire for JSON logs, optional verbose logging, and centralized audit logging of command execution, but I respectfully disagree with how this logic is implemented in the util subdirectory, and I do not want us to continue using it.

- Logging levels should be configured and handled using standard Python logging configuration. Standard Python loggers already support JSON logging with no need for further modification.
- Aside from configuring logging when the process starts, there should not be any custom logging functions imported and used around the codebase. Standard Python logging already provides all the capabilities necessary for that.
- I don't think we should be using centralized helpers to execute commands; instead, we should use standard recommended Python methods for safely executing code, and audit our code to check that no string interpolation happens before strings are passed to the shell (there is no need to interpolate command arguments in our current system).
- I recognize the desire to log all command execution in a centralized manner. Although we could use a thin helper wrapper around `subprocess.run` to log arguments agnostically, I think the best solution is the system auditing framework provided in Python 3.8:
```python
import sys, subprocess, logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)
def handle_audit_event(event, details):
    if event == "subprocess.Popen":
        logger.info({"event": "popen", "details": details})
sys.addaudithook(handle_audit_event)
subprocess.run(['ls', '/'])
popen ('ls', ['ls', '/'], None, None)
```